diff -Nru upstream/pkg/filepath/filepath.cabal boot/pkg/filepath/filepath.cabal
--- upstream/pkg/filepath/filepath.cabal	2018-01-23 23:17:57.000000000 +0100
+++ boot/pkg/filepath/filepath.cabal	2018-01-23 23:17:57.000000000 +0100
@@ -47,6 +47,10 @@
         System.FilePath.Posix
         System.FilePath.Windows
 
+    if impl(GHCJS >= 0.1)
+        Exposed-modules:
+            System.FilePath.Current
+
     build-depends:
         base >= 4 && < 4.12
 
diff -Nru upstream/pkg/filepath/System/FilePath/Current.hs boot/pkg/filepath/System/FilePath/Current.hs
--- upstream/pkg/filepath/System/FilePath/Current.hs	1970-01-01 01:00:00.000000000 +0100
+++ boot/pkg/filepath/System/FilePath/Current.hs	2018-01-22 18:25:10.000000000 +0100
@@ -0,0 +1,4 @@
+{-# LANGUAGE CPP #-}
+#define MODULE_NAME     Current
+#define IS_CURRENT      1
+#include "Internal.hs"
diff -Nru upstream/pkg/filepath/System/FilePath/Internal.hs boot/pkg/filepath/System/FilePath/Internal.hs
--- upstream/pkg/filepath/System/FilePath/Internal.hs	2018-01-23 23:17:57.000000000 +0100
+++ boot/pkg/filepath/System/FilePath/Internal.hs	2018-01-23 23:17:59.000000000 +0100
@@ -1,4 +1,6 @@
-#if __GLASGOW_HASKELL__ >= 704
+#ifdef IS_CURRENT
+{-# LANGUAGE ForeignFunctionInterface, JavaScriptFFI, Trustworthy #-}
+#elif __GLASGOW_HASKELL__ >= 704
 {-# LANGUAGE Safe #-}
 #endif
 {-# LANGUAGE PatternGuards #-}
@@ -63,6 +65,8 @@
 -- [1] <http://msdn.microsoft.com/en-us/library/windows/desktop/aa365247.aspx Naming Files, Paths and Namespaces> (Microsoft MSDN)
 module System.FilePath.MODULE_NAME
     (
+    isWindows, isPosix,
+
     -- * Separator predicates
     FilePath,
     pathSeparator, pathSeparators, isPathSeparator,
@@ -126,8 +130,12 @@
 
 -- | Is the operating system Windows like
 isWindows :: Bool
+#ifdef IS_CURRENT
+isWindows = js_isWindows
+foreign import javascript unsafe "h$filepath_isWindows()" js_isWindows :: Bool
+#else
 isWindows = IS_WINDOWS
-
+#endif
 
 ---------------------------------------------------------------------
 -- The basic functions
